### ISSUE TITLE
Unsupported keras layer type Lambda

### ISSUE THREAD

**Issue by saudet** [deeplearning4j/deeplearning4j#4256]
#### Issue Description

While trying to import a Keras model for YOLO created via YAD2K, I get an "Unsupported keras layer type Lambda" error for this layer:
```
space_to_depth_x2 (Lambda)       (None, 19, 19, 256)   0           leaky_re_lu_21[0][0]     
```
It corresponds to https://www.tensorflow.org/api_docs/python/tf/space_to_depth in TensorFlow. It does not look like a complicated operation, but do you know if we have a layer like that somewhere in DL4J? What do you think  @AlexDBlack?

#### Version Information

* Deeplearning4j 0.9.2-SNAPSHOT

**Comment by maxpumperla** [comment]
@saudet lambda layers in keras allow you to essentially provide a little lambda function of your choice, a simple way to add a new layer. Supporting `Lambda` in general in DL4J would need Samediff in place, as we need to compute the backward pass. So, I think it will take a while for us to fully support this.

On the other hand, the specific operation you mention, sure, we could implement that as a one-off. How urgent or important is this?

**Comment by saudet** [comment]
We're going to need it for accuracy with object detection at some point. I think @wangfeng-skymind might need it soon (?) so it may be semi-urgent... Do we have a layer like space_to_depth already though?

**Comment by maxpumperla** [comment]
@raver119 is currently implementing a ton of things for tf import. if we have an equivalent of `space_to_depth` in libnd4j/nd4j, I could relatively quickly build a dl4j layer around it.

**Comment by raver119** [comment]
space_to_depth is implemented.

**Comment by maxpumperla** [comment]
@saudet we even have this in SameDiff by now. Supporting lambda layers in general is still a long way to go, but we could have this a utility layer. Is this still important or urgent or both? :)

**Comment by nlebreton** [comment]
@maxpumperla thx for this impl. I will test it by trying to import a (full) yolo v2 asap and update this issue accordingly.

**Comment by maxpumperla** [comment]
@nlebreton wait, it's in ND4J right now, I still need to wrap this into DL4J _and_ support keras import.

@saudet can you please (again, I guess) point me to the model specification that you try to import?

**Comment by saudet** [comment]
IMO, I think it's important to be able to import full YOLOv2 models for decent accuracy with object detection. I'll leave it up to @jpatanooga  and @agibsonccc to set priorities about this and other things though.

**Comment by nlebreton** [comment]
I can confirm that from a business point of view, Tiny Yolo is not accurate enough for many use-cases. I work on a proof of concept trying to convince my partners that java and dl4j are viable compared to the classical python approach. One strong argument is "Keras can be imported" but today state of the art objects detection model Yolo cannot be imported. I really hope that this capability will be added soon.

**Comment by maxpumperla** [comment]
@nlebreton @saudet we're slowly but surely getting there, see: https://github.com/deeplearning4j/deeplearning4j/pull/4691

the full yolo h5 exceeds the github file size limit (it's well over 200mb), so I can't just put it into our resources repo. Is there a smaller version that reproduces the basic pattern (i.e. concat a conv layer with the output of space_to_depth)?

**Comment by nlebreton** [comment]
@maxpumperla I was able to reproduce the issue with that one ? Its size is 36kb.
[my_model.zip](https://github.com/deeplearning4j/deeplearning4j/files/1749342/my_model.zip)

**Comment by maxpumperla** [comment]
<img width="639" alt="screen shot 2018-02-23 at 15 59 17" src="https://user-images.githubusercontent.com/3462566/36600458-9f7fec76-18b2-11e8-82a7-b29905c9a6d6.png">

@nlebreton @saudet importing full `yolo-voc` works on the branch now, will try to test some more, then wrap this up.

**Comment by maxpumperla** [comment]
perfect, `yolo` seems to work as well. however, I see that your `my_model.zip` fails for some reason @nlebreton.

**Comment by maxpumperla** [comment]
ok, I also added a keras configuration test (loading json only), which you can see here: 

https://github.com/deeplearning4j/deeplearning4j/blob/4a25300f1452d656b15686cbb224f28c28aa70c1/deeplearning4j-modelimport/src/test/java/org/deeplearning4j/nn/modelimport/keras/configurations/Keras2ModelConfigurationTest.java#L44-L46

**Comment by maxpumperla** [comment]
I'm renaming this issue to this more specific case and open a new "Lambda layer support" issue. thanks!
